lib/elixir/lib/kernel.ex

```diff
@@ -4998,6 +4998,56 @@ defmodule Kernel do
     Macro.escape(NaiveDateTime.from_iso8601!(string))
   end
 
+  @doc ~S"""
+  Handles the sigil `~U` to create a UTC `DateTime`.
+
+  The lower case `~u` variant does not exist as interpolation
+  and escape characters are not useful for datetime sigils.
+
+  The given `datetime_string` must include "Z" or "00:00" offset which marks it
+  as UTC, otherwise an error is raised.
+
+  More information on date times can be found in the `DateTime` module.
+
+  ## Examples
+
+      iex> ~U[2015-01-13 13:00:07Z]
+      ~U[2015-01-13 13:00:07Z]
+      iex> ~U[2015-01-13T13:00:07.001+00:00]
+      ~U[2015-01-13 13:00:07.001Z]
+
+  """
+  @doc since: "1.9.0"
+  defmacro sigil_U(datetime_string, modifiers)
+
+  defmacro sigil_U({:<<>>, _, [string]}, []) do
+    Macro.escape(datetime_from_utc_iso8601!(string))
```

Comment:
What do you think about?

  * having a different error message that expresses that the `~U` sigil only works with "Z" or "00:00". Besides, `:invalid_offset` might mean that the offset is invalid regarding format, for example `2015-01-23T23:50:07,123+52:30` (that actually returns `:invalid_format`). 
  * collapsing the logic into a single function

```elixir
defmacro sigil_U({:<<>>, _, [string]}, []) do
  case DateTime.from_iso8601(string) do
    {:ok, utc_datetime, 0} ->
      Macro.escape(utc_datetime)

    {:ok, _datetime, _offset} ->
      raise ArgumentError,
            "~U sigil expected a UTC date time ("Z" or "00:00" offset), got: #{inspect(string)}"

    {:error, reason} ->
      raise ArgumentError,
            "cannot parse #{inspect(string)} as UTC date time, reason: #{inspect(reason)}"
  end
end
```

Reply:
> * Besides, `:invalid_offset` might mean (...)

Do you mean that users might interpret `:invalid_offset` as something other than it is right now? And yeah, if offset is not formatted properly we get `:invalid_format`, if it's not there we get `:missing_offset`. If `:invalid_offset` is too generic, maybe it should be more specific e.g. `:non_utc_offset` or something along the lines?

Otherwise, I think your suggestion about error message is a good one too.

Reply:
Regarding collapsing everything into a single macro, I followed style of existing sigils that the macro does as little as possible (I quite like this approach in general too), and besides there's compilation issue when I tried your suggestion:

```
== Compilation error in file lib/kernel.ex ==
** (UndefinedFunctionError) function Kernel.unescape_tokens/1 is undefined (function not available)
    (elixir) Kernel.unescape_tokens/1
    (stdlib) erl_eval.erl:567: :erl_eval.local_func/6
    (stdlib) erl_eval.erl:888: :erl_eval.expr_list/6
    (stdlib) erl_eval.erl:240: :erl_eval.expr/5
    (elixir) expanding macro: Kernel.sigil_s/2
    (elixir) lib/kernel.ex:5028: Kernel.sigil_U/2
    (elixir) expanding macro: Kernel.raise/2
    (elixir) lib/kernel.ex:5027: Kernel.sigil_U/2
make: *** [lib/elixir/ebin/Elixir.Kernel.beam] Error 1
```

I'd be OK with collapsing `datetime_from_iso_8601` into `datetime_from_iso_8601!`  :)

Reply:
`:non_utc_offset` sounds good 👍

> there's compilation issue when I tried your suggestion:

Sorry, I wrote that on the go in my phone and I could not check it before posting.

> I'd be OK with collapsing `datetime_from_iso_8601` into `datetime_from_iso_8601!` :)

Perfect 👌